File: #17_Letter_Combinations_of_a_Phone_Number/Solution.java
/* dfs */

import java.util.*;

public class Solution {
	static String[] map = {"abc", "def", "ghi", "jkl", "mno", "pqrs", "tuv", "wxyz"};
    public List<String> letterCombinations(String digits) {
    	List<String> result = new LinkedList();
    	if (digits == null || digits.length() == 0) return result;
    	StringBuilder sb = new StringBuilder();
    	bt(digits, sb, result);
    	return result;
    }

    private void bt(String s, StringBuilder sb, List<String> result) {
    	if (s == null || s.length() == 0) return;
    	int num = s.charAt(0) - '0';
    	if (num < 2) return;
    	for (char c : map[num - 2].toCharArray()) {
    		sb.append(c);
    		if (s.length() == 1) result.add(sb.toString());
    		bt(s.substring(1, s.length()), sb, result);
    		sb.deleteCharAt(sb.length() - 1);
    	}
    }

    public static void main(String[] args) {
    	Solution sol = new Solution();
    	List<String> re = sol.letterCombinations("23");
    	for(String s : re) {
    		System.out.println(s);
    	}
    }
}

File: #17_Letter_Combinations_of_a_Phone_Number/Solution.py
'''
Backtracing DFS solution
'''

# class Solution:
# 	# @param {string} digits
# 	# @return {string[]}
# 	def letterCombinations(self, digits):
# 		if not digits:
# 			return []
# 		map = {'0': '',
# 				'1':    '', '2': 'abc', '3': 'def', 
# 				'4': 'ghi', '5': 'jkl', '6': 'mno', 
# 				'7': 'pqrs', '8': 'tuv', '9': 'wxyz'}
# 		re = []
# 		self.dfs(re, map, digits, '')
# 		return re

# 	def dfs(self, result, map, digits, str):
# 		if not digits:
# 			result.append(str)
# 			return
# 		for c in map[digits[0]]:
# 			self.dfs(result, map, digits[1:], str + c)

'''
BFS soltuion
'''
from collections import deque

class Solution:
	# @param {string} digits
	# @return {string[]}
	def letterCombinations(self, digits):
		if not digits:
			return []
		map = {'0': '',
				'1':    '', '2': 'abc', '3': 'def', 
				'4': 'ghi', '5': 'jkl', '6': 'mno', 
				'7': 'pqrs', '8': 'tuv', '9': 'wxyz'}
		queue = deque([''])
		for index, key in enumerate(digits):
			while len(queue[0]) == index:
				str = queue.popleft()
				for c in map[key]:
					queue.append(str + c)
		return queue


sol = Solution()
re = sol.letterCombinations("23")
for str in re:
	print str,
